### PR TITLE
TASK: Check Neos.Demo refactoring for ES CR specifics

### PR DESCRIPTION
**Pull request by ahaeslich**
- New dimension configuration format incl. dimension value mapping to url path segment
- Use Neos.Caching.entryIdentifierForNode() helper in Fusion
- Check usage of `Neos.Node.inBackend(node)` instead of `node.context.inBackend`
- Use `Neos.Dimension` helper to render language menu in refactored header

### CODE REVIEW

**Comment by mhsdesign** [comment]
help

File: Resources/Private/Fusion/Integration/Header.fusion [hunk 1]
```diff
@@ -18,7 +18,13 @@ prototype(Neos.Demo:Integration.Header.Renderer) < prototype(Neos.Demo:Presentat
         maximumLevels = 2
     }
 
-    currentLanguage = ${site.context.dimensions.language[0]}
+    currentLanguage = ${Neos.Dimension.currentValue(site, 'language')}
+    languages = Neos.Fusion:Map {
+        items = ${Neos.Dimension.all(site)['language'].values.iterator}
```
Comment (ahaeslich) [inline Resources/Private/Fusion/Integration/Header.fusion]: This works but there are two issues I have with this line of code:

- In neos/neos-development-collection#4085 we removed the `findPotentialDimensionValues` function but this is now the second time of the usage `${Neos.Dimension.all(site)['language']}` so I guess we should reintroduce it. Maybe with a better naming ;)
- If we have a collection object at hand (`values` is of type `ContentDimensionValues`) do we want to see this usage of `iterator` to work with a e.g. `Neos.Fusion:Map` or similar Implementations which as of now only accept an `array`.

Any thoughts @skurfuerst / @bwaidelich / @mficzel ?

Comment (mficzel) [reply Resources/Private/Fusion/Integration/Header.fusion]: 1. I also think we will introduce a permutation of "findPotentialDimensionValues" (all values) again eventually. Probably in distinction to "findDimensionRootValue" (values without fallback). Better names would really help.
2. We should make `Map` accept an `iterator` that does not implement `countable`. The only value we cannot calculate in this case is `iterator.isLast`.

**Comment by ahaeslich** [comment]
> help

Better @mhsdesign 😉?

File: Resources/Private/Fusion/Integration/Header.fusion (hunk 1 above)
Comment (bwaidelich) [reply Resources/Private/Fusion/Integration/Header.fusion]: I didn't look into it very closely, but IMO this Fusion code is _far_ too low level.
Couldn't we add a dedicated helper, like
```
languages = ${Neos.Dimension.nameToBeFound(site, 'language')}
```
?

Comment (mficzel) [reply Resources/Private/Fusion/Integration/Header.fusion]: Looping/Mapping over uncountable collections seems to be rather simple see https://github.com/neos/neos-development-collection/pull/4168 ... however making collections countable would often make sense.

Comment (skurfuerst) [reply Resources/Private/Fusion/Integration/Header.fusion]: (just discussing with @ahaeslich )

suggestion / idea:

```
// "language" would be ContentDimensionId
languages = ${Neos.Dimension.allDimensionValues(site, 'language')}
```

-> should return `ContentDimensionValues` object I'd say -> this is iterable.

Comment (bwaidelich) [reply Resources/Private/Fusion/Integration/Header.fusion]: @skurfuerst yes, I like that a lot!

Comment (ahaeslich) [reply Resources/Private/Fusion/Integration/Header.fusion]: Will add that so we can proceed with this MR and Neos.Seo

**Review by kdambekalns — APPROVED** [review]
✅ by 👀

**Pull request merged**